### PR TITLE
Fix nil pointer dereferencing when converting `vars` to `replacements `

### PR DESCRIPTION
**Pull request by Aceralon**
Fix `patch.Target` is nil in `writePatchTargets`

https://github.com/kubernetes-sigs/kustomize/blob/9e42f8d57e285d1113d608895a2015cea9ad2b21/kustomize/commands/edit/fix/convert.go#L284-L287

When converting a kustomization file that contains `Patch` with only `Path` field, using `--vars` might cause kustomize to panic for dereferencing nil pointer.

Example for the bug:
```yaml
#kustomization.yaml
patchesStrategicMerge:
  - patch.yaml

vars:
  - name: CERTIFICATE_NAMESPACE
    objref:
      name: system
    fieldref:
      fieldpath: metadata.namespace
```

```yaml
#patch.yaml
apiVersion: apps/v1
kapiVersion: apps/v1
kind: Deployment
metadata:
  name: controller-manager
  namespace: system
spec:
  template:
    spec:
      containers:
        - name: $(CERTIFICATE_NAMESPACE)
```
Then run `kustomise edit fix --vars` under this folder.

### CODE REVIEW

File: kustomize/commands/edit/fix/convert.go [hunk 1]
```diff
@@ -134,7 +135,7 @@ func getNodesFromFile(fileName string, fSys filesys.FileSystem) ([]*kyaml.RNode,
 	}
 	out := &bytes.Buffer{}
 	r := kio.ByteReadWriter{
-		Reader:                bytes.NewBufferString(string(b)),
+		Reader:                bytes.NewReader(b),
```
Comment (Aisuko) [inline kustomize/commands/edit/fix/convert.go:138]: Please add a code comment here that the reason we have to change to `NewReader()`

**Review by Aisuko — CHANGES_REQUESTED** [review]
Thanks for your contribution, we do not have any test cases covering this function `ConvertVarsToReplacements`, nor `writePatchTargets`. To make us more confident to merge this PR, would you mind add a test case for this PR into [convert_test.go](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/edit/fix/convert_test.go)?

File: kustomize/commands/edit/fix/convert.go (hunk 1 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert.go:138]: The `NewReader()` can directly create `io.Reader` from `[]byte`, I think it is redundant to convert a `[]byte` to `string` then back to `[]byte`. 

Internally, the `NewBufferString` will just convert `string` back to `[]byte`.

```golang
func NewBuffer(buf []byte) *Buffer { return &Buffer{buf: buf} }

func NewBufferString(s string) *Buffer {
	return &Buffer{buf: []byte(s)}
}
```

**Comment by Aceralon** [comment]
> Thanks for your contribution, we do not have any test cases covering this function `ConvertVarsToReplacements`, nor `writePatchTargets`. To make us more confident to merge this PR, would you mind add a test case for this PR into [convert_test.go](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/edit/fix/convert_test.go)?

Added the test case.

File: kustomize/commands/edit/fix/convert.go (hunk 1 above)
Comment (Aisuko) [reply kustomize/commands/edit/fix/convert.go:138]: I got your point; I have the same idea here for the code, but we cannot see more useful from `NewReader(b)`, I want to make sure it will not cause any other issues. So here we need to add the testcase.

Comment (Aceralon) [reply kustomize/commands/edit/fix/convert.go:138]: Ok. So do I need to add test cases directly for `getNodesFromFile`?

I've add a test case for the nil pointer dereferencing including `writePatchTargets`

**Comment by Aisuko** [comment]
> > Thanks for your contribution, we do not have any test cases covering this function `ConvertVarsToReplacements`, nor `writePatchTargets`. To make us more confident to merge this PR, would you mind add a test case for this PR into [convert_test.go](https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/edit/fix/convert_test.go)?
> 
> Added the test case.

Thank you.

**Comment by Aisuko** [comment]
cc @koba1t may you please look at this one? And we need to trigger the CI process.

**Comment by koba1t** [comment]
@Aisuko 
Sorry, I don't have that permission to run Github Actions...
Please wait for a check from the maintainer.

/cc @annasong20 @natasha41575

**Comment by Aisuko** [comment]
> @Aisuko Sorry, I don't have that permission to run Github Actions... Please wait for a check from the maintainer.
> 
> /cc @annasong20 @natasha41575

Got it, thank you.

**Comment by Aceralon** [comment]
Rebased to the latest master.

**Comment by natasha41575** [comment]
/ok-to-test

@Aceralon Apologies, could you rebase again to trigger the test suite?

**Comment by Aceralon** [comment]
> /ok-to-test
> 
> @Aceralon Apologies, could you rebase again to trigger the test suite?

Rebased to the master.

**Review by ncapps — COMMENTED** [review]
I am not an official reviewer but this change lgtm.

File: kustomize/commands/edit/fix/convert.go [hunk 2]
```diff
@@ -10,12 +10,13 @@ import (
 	"strconv"
 	"strings"
 
+	"sigs.k8s.io/yaml"
+
```
Comment (natasha41575) [inline kustomize/commands/edit/fix/convert.go:14]: nit: remove the newline

File: kustomize/commands/edit/fix/convert_test.go [hunk 3]
```diff
@@ -1143,3 +1144,76 @@ metadata:
...
+spec:
+  template:
+    spec:
+      containers:
+        - name: $(CERTIFICATE_NAMESPACE)
+`)
+
+	fSys := filesys.MakeFsInMemory()
+	testutils_test.WriteTestKustomizationWith(fSys, kustomization)
+	fSys.WriteFile("patch.yaml", patch)
```
Comment (natasha41575) [inline kustomize/commands/edit/fix/convert_test.go:1179]: lint error: `Error return value of `fSys.WriteFile` is not checked (errcheck)`

```suggestion
	assert.NoError(fSys.WriteFile("patch.yaml", patch))
```

File: kustomize/commands/edit/fix/convert.go (hunk 1 above)
Comment (natasha41575) [inline kustomize/commands/edit/fix/convert.go:138]:
```suggestion
		Reader:                bytes.NewBuffer(b),
```

In https://github.com/kubernetes-sigs/kustomize/pull/5211/files#r1247400428 you explain that we don't want to convert to string and back to byte, which is fine, but it is not clear why we need to switch from `Buffer` to `Reader`. 

If there is some reason to switch to `Reader`, I would prefer that happen in its own PR with a description, to avoid conflating it with the nil pointer error change.

**Review by natasha41575 — APPROVED** [review]
Couple of very small comments. Overall LGTM

**Review by natasha41575 — CHANGES_REQUESTED** [review]
accidentally approved instead of requesting changes, whoops

File: kustomize/commands/edit/fix/convert.go (hunk 2 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert.go]: removed

File: kustomize/commands/edit/fix/convert_test.go (hunk 3 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert_test.go]: added the assert for all `fSys.WriteFile`

File: kustomize/commands/edit/fix/convert.go (hunk 1 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert.go]: Using `NewBuffer` instead

File: kustomize/commands/edit/fix/convert_test.go [hunk 4]
```diff
@@ -1143,3 +1143,76 @@ metadata:
...
+vars:
+  - name: CERTIFICATE_NAMESPACE
+    objref:
+      name: system
+    fieldref:
+      fieldpath: metadata.namespace
+`)
+	patch := []byte(`
+apiVersion: apps/v1
+kapiVersion: apps/v1
```
Comment (stormqueen1990) [inline kustomize/commands/edit/fix/convert_test.go:1164]: I believe this might be a typo (repeated on lines 1206-1207).

File: kustomize/commands/edit/fix/convert_test.go [hunk 5]
```diff
@@ -1143,3 +1143,76 @@ metadata:
...
+	assert.NoError(t, cmd.Flags().Set("vars", "true"))
+	assert.NoError(t, cmd.RunE(cmd, nil))
+	content, err := testutils_test.ReadTestKustomization(fSys)
+	assert.NoError(t, err)
+
+	assert.Equal(t, `
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+
```
Comment (stormqueen1990) [inline kustomize/commands/edit/fix/convert_test.go:1189]: Are these empty lines meant to be here?

**Review by stormqueen1990 — COMMENTED** [review]
Mostly LGTM, but the tests seem to have a small typo.

File: kustomize/commands/edit/fix/convert_test.go (hunk 4 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert_test.go]: Removed the second line.

File: kustomize/commands/edit/fix/convert_test.go (hunk 5 above)
Comment (Aceralon) [reply kustomize/commands/edit/fix/convert_test.go:1188]: Yes, as there is an empty line between L1150 and L1152

Comment (stormqueen1990) [reply kustomize/commands/edit/fix/convert_test.go:1188]: I see, makes sense!

**Comment by Aceralon** [comment]
> Couple of very small comments. Overall LGTM

@natasha41575 Finished all the changes.

**Comment by natasha41575** [comment]
/lgtm

Sorry, missed adding the lgtm tag with my approval
